File: typescript/src/external_task_bpmn_error.ts
import {IIdentity} from '@essential-projects/iam_contracts';

import {IExternalTaskApi, IExternalTaskResult} from '.';

/**
 * Contains the result set for an ExternalTask that failed with an Error.
 */
export class ExternalTaskBpmnError implements IExternalTaskResult {

  private readonly externalTaskId: string;
  private readonly errorCode: string;

  constructor(externalTaskId: string, errorCode: string) {
    this.externalTaskId = externalTaskId;
    this.errorCode = errorCode;
  }

  /**
   * Sends the ExternalTasks result to the ExternalTaskAPI, using the given 'IExternalTaskAPI' instance.
   *
   * @async
   * @param externalTaskApi    The ExternalTaskAPI instance that should process the ExternalTasks result.
   * @param identity           The identity to use for sending the ExternalTasks result.
   * @param workerId           Id of the Worker wich handled this tasks.
   */
  public async sendToExternalTaskApi(externalTaskApi: IExternalTaskApi, identity: IIdentity, workerId: string): Promise<void> {
    await externalTaskApi.handleBpmnError(identity, workerId, this.externalTaskId, this.errorCode);
  }

}

File: dotnet/src/ExternalTaskBpmnError.cs
namespace ProcessEngine.ExternalTaskAPI.Contracts
{
    using System;
    using System.Threading.Tasks;

    using EssentialProjects.IAM.Contracts;

    /// <summary>
    /// Contains the result set for a successfully executed ExternalTask.
    /// </summary>
    public class ExternalTaskBpmnError : IExternalTaskResult
    {
        private readonly string externalTaskId;
        private readonly string errorCode;

        /// <summary>
        /// Creates a new Instace of ExternalTaskBpmnError
        /// </summary>
        /// <param name="externalTaskId">
        /// Id of the ExternalTask
        /// </param>
        /// <param name="errorCode">
        /// ErrorCode of executing ExternalTask
        /// </param>
        public ExternalTaskBpmnError(string externalTaskId, string errorCode)
        {
            this.externalTaskId = externalTaskId;
            this.errorCode = errorCode;
        }

        /// <summary>
        /// Sends the ExternalTasks result to the ExternalTaskAPI, using the given 'IExternalTaskAPI' instance.
        /// </summary>
        /// <param name="externalTaskAPI">
        /// The ExternalTaskAPI instance that should process the ExternalTasks result.
        /// </param>
        /// <param name="identity">
        /// The identity to use for sending the ExternalTasks result.
        /// </param>
        /// <param name="workerId">
        /// Id of the Worker wich handled this tasks.
        /// </param>
        public async Task SendToExternalTaskApi(IExternalTaskAPI externalTaskAPI, IIdentity identity, string workerId)
        {
            await externalTaskAPI.HandleBpmnError(identity, workerId, this.externalTaskId, this.errorCode);
        }
    }
}
